rs/nns/governance/src/proposals/mod.rs

```diff
@@ -54,3 +59,82 @@ pub(crate) fn invalid_proposal_error(reason: &str) -> GovernanceError {
+///
+/// For example, suppose this returns ({42 => 3.14}, 99.9). This means that
+/// neuron 42 should get 3.14/99.9 of today's reward purse.
+///
+/// Non-essential fact: It is highly unusual for result.1 to be the sum of the
+/// values in result.0. The main reason they would not be equal is that not all
+/// neurons vote. Another (less imporant reason) is that some neurons lose
+/// voting power due to inactivity.
+pub fn sum_weighted_voting_power<'a>(
+    proposals: impl Iterator<Item = (ProposalId, Option<&'a ProposalData>)>,
+) -> (HashMap<NeuronId, /* exercised */ f64>, /* total */ f64) {
+    // Results.
+    let mut neuron_id_to_exercised_weighted_voting_power: HashMap<NeuronId, f64> = HashMap::new();
+    let mut total_weighted_voting_power = 0.0;
+
+    for (proposal_id, proposal) in proposals {
+        let proposal = match proposal {
+            Some(ok) => ok,
+            None => {
+                println!(
+                    "{}ERROR: Trying to give voting rewards for proposal {}, \
+                     but it was not found.",
+                    LOG_PREFIX, proposal_id.id,
+                );
+                continue;
+            }
+        };
+
+        let reward_weight = proposal.topic().reward_weight();
+
+        // This is used in lieu of total_potential_voting_power. That is, this
+        // gets used if proposal does not have total_potential_voting_power.
+        // This fall back will only be used during a short transition period
+        // when there is a backlog of "old" proposals that were created without
+        // total_potential_voting_power, but all new proposals have it. In the
+        // case of such legacy proposals, their total potential voting power
+        // would have been equal to this anyway, so this is a sound substitute
+        // for total_potential_voting_power.
+        let mut total_ballots_voting_power = 0;
+
+        for (neuron_id, ballot) in &proposal.ballots {
+            total_ballots_voting_power += ballot.voting_power;
+
+            // Don't reward neurons that did not actually vote. (Whereas, ALL
+            // eligible neuron gets an "empty" ballot when the proposal is first
+            // created. An "empty" ballot is one where the vote field is set to
+            // Unspecified.)
+            let vote = Vote::try_from(ballot.vote).unwrap_or_else(|err| {
+                println!(
+                    "{}ERROR: Unrecognized Vote {} in ballot by \
+                         neuron {} on proposal {:?}: {:?}",
+                    LOG_PREFIX, ballot.vote, neuron_id, proposal.id, err,
+                );
+                Vote::Unspecified
+            });
+            if !vote.eligible_for_rewards() {
+                continue;
+            }
+
+            // Increment neuron.
+            *neuron_id_to_exercised_weighted_voting_power
+                .entry(NeuronId { id: *neuron_id })
+                .or_insert(0.0) += (ballot.voting_power as f64) * reward_weight;
```

Comment:
Is there a good enough reason why we're using floating point for reward calculation instead of `Decimal`?